### ISSUE TITLE
[Planning Bug] Apollo cannot reasonably avoid large oncoming vehicles

### ISSUE THREAD

**Issue by AV-Security** [ApolloAuto/apollo#14628]
**Describe the bug**

The body of a large NPC on the adjacent lane partially encroached on the ego's lane, there was room for the ego to evade to the side, but it stopped in place and collided with the NPC.
*Note: Simulator platform was LGSVL and the perception results were ground-truth data provided by LGSVL.

**System information**
OS Platform and Distribution: Ubuntu 18.04
Apollo installed from (source or binary): source
Apollo version : 6.0 (master branch - commit b5bd74112e2ec5d7c65cad6bea592353de566156)

**To Reproduce**

The video corresponding to the collision scenario is in the attachment.

**Expected behavior**

The ego should evade to the side when the NPC were approaching.

**Additional context**

It seems that `PathBoundsDecider` does not take into account the predicted trajectory of dynamic obstacles when calculating the boundaries of the drivable road, resulting in the planned path may be blocked by dynamic obstacles.

**Demo Video**

https://user-images.githubusercontent.com/87695978/193569751-6e9ed196-1d05-493f-b28a-24d7ff6f9d8e.mp4

**Attachments**
[Collision Video](https://drive.google.com/file/d/1yiut_06NN3qsbMrLAoGFAbmJuEOk6u95/view?usp=sharing)

**Comment by ntutangyun** [comment]
You might be asking a little bit too much of it. 

As I mentioned in the another issue, this open source version of Apollo is still under rapid development and it is expected to have a lot of corner cases such as this one not handled. 

But upon encountering such a scenario, Apollo (as shown in the video) did its best as it slowly comes to a full stop (waiting for the truck to pass or for a manual takeover fallback). 

I would rather argue that the collision is mainly caused by the "lane centering" behavior of the truck NPC controlled by the simulator, which if you think about it, also does not implement a mechanism to handle this type of scenario. 

In addition, the default configuration for the operational lane width is about 3.5 meters I believe. So one may also argue that you are already driving in an environment (with way too narrow lanes) out of Apollo's Operational Design Domain (ODD). So again I believe Apollo did its best I think...

**Comment by YuqiHuai** [comment]
@ntutangyun very detailed response! I have a follow question regarding this, how did you know the default operational lane width is 3.5 meters? and is Apollo's OOD formally defined somewhere? So far https://daobook.github.io/apollo/README.html seems to be the most detailed documentation I found.

**Comment by ntutangyun** [comment]
Sometimes you can rely on the apollo documents. but sometimes you can't..because the document may be too obsolete. 

For the default lane width, there is a configuration parameter in apollos planning module. but I forgot where in detail... you may be able to find it by searching the keywords such as `lane_width` or `laneWidth`. hope it helps.

**Comment by YuqiHuai** [comment]
Thanks! @ntutangyun I will try to find the value in I guess the configuration files for the planning module.

**Comment by AV-Security** [comment]
@ntutangyun thanks for your explanation. In my opinion, even with this increase in road width, there are still exits vehicles taking up part of the lane, for example in life some cars do not follow the centerline of the lane strictly. What I care about is that Apollo has the space to evade to the right safely but doesn’t do so. So it may back to the limitation that Apollo's ways of avoiding dynamic obstacles (slowing down or stopping) are too simple to safely handle some complex emergencies like this.

**Comment by ntutangyun** [comment]
By examing the video closely I just noticed that the HD map itself may be incorrect... apparently, the truck's lane and apollo's lane are overlapping... could you fix this issue and try again?

<img width="1262" alt="Screenshot 2022-10-08 at 12 29 05 PM" src="https://user-images.githubusercontent.com/16061105/194688206-0c1c801f-7d9a-4430-a6f4-7310d4d09e59.png">

**Comment by AV-Security** [comment]
@ntutangyun Thank you for pointing out the problem with this map. I will try to fix the map later.
I found another example where the lanes did not overlap with each other. The ego did not reasonably evade the NPC whose body partially occupied the ego’s lane as well.

https://user-images.githubusercontent.com/87695978/194744291-801392fa-c734-4c38-b6da-9bade36fe1ff.mp4


In our test, we also found that if NPC is directly fixed at the collision position, the ego would drive to the right appropriately to avoid this NPC.

https://user-images.githubusercontent.com/87695978/194744304-face7ba8-a82e-4426-8816-c2bc3dea6784.mp4

In order to investigate this, I analyzed the code of the planning module and found that the function `PathBoundsDecider:: GenerateRegularPathBound` in `PathBoundsDecider` is used to calculate the drivable road boundary of ego. As shown in the following code, the Apollo only considers static obstacles and ignores adjusting its drivable boundary according to dynamic obstacles.
 - commit id:b5bd74112e2ec5d7c65cad6bea592353de566156
- modules/planning/tasks/path_bounds_decider/path_bounds_decider.cc::368
- PathBoundsDecider::GenerateRegularPathBound

<img width="613" alt="截屏2022-10-09 15 27 59" src="https://user-images.githubusercontent.com/87695978/194744461-5a9e8d1f-e1f7-46e6-8083-5ca49bbe4773.png">

So I think it would be safer to adjust the ego’s drivable boundary according to the predicted trajectories of dynamic obstacles.

**Comment by ntutangyun** [comment]
Thanks for the detailed investigation. So basically it's a feature to be implemented in the future.
